### PR TITLE
fix(pgvector): replace lone surrogates so a transcript surrogate no longer aborts the mine (#1833)

### LINKED ISSUE #1833: pgvector backend aborts the whole mine run on a lone surrogate (U+D800-U+DFFF) in transcript content
# Bug: pgvector backend aborts the whole mine run on a lone surrogate (U+D800-U+DFFF) in transcript content

**Component:** `mempalace/backends/pgvector.py` (`_PgVectorClient.upsert_rows`)
**Affected version:** develop @ `94d1ced`; same `upsert_rows` write path as #1829
**Severity:** Medium-High. A single offending input file aborts the entire mine; content that ChromaDB ingests fine cannot be mined under pgvector.
**Affected backends:** pgvector and sqlite_exact abort (verified end to end). ChromaDB (the default) strips surrogates from documents only, so it tolerates a document-surrogate but still aborts on a metadata-surrogate. Not pgvector-specific, unlike the NUL bug in #1829.

## Summary

This is the lone-surrogate sibling of #1829. `upsert_rows()` passes `document` (a `text`
column) and `metadata` (a `jsonb` column) straight to Postgres via psycopg. A lone UTF-16
surrogate (U+D800 to U+DFFF) has no UTF-8 encoding, so psycopg cannot encode the parameter
and raises, aborting the whole `mine` run. Lone surrogates reach transcript content the same
way NUL does (captured tool output, `surrogateescape`-decoded bytes), and are known to occur
in real MCP content.

## Symptom

```
mempalace.backends.base.BackendError: pgvector query failed:
'utf-8' codec can't encode character '\ud800' in position 6: surrogates not allowed
```

The mine exits non-zero mid-run; every file after the offending one in that run is left
unmined. Both `document` and `metadata` raise the same error (unlike #1829, where the text
and jsonb columns failed with two different messages).

## Root cause

`_PgVectorClient.upsert_rows()` builds the parameter tuples directly from `row["id"]`,
`row["document"]`, and `_json_dumps(row.get("metadata"))` with no sanitization, then
`INSERT ... VALUES (%s, %s, %s::jsonb, ...)`. The bulk mine path does not go through the
per-drawer sanitize path (which already strips lone surrogates via
`config.strip_lone_surrogates`, `config.py`), so nothing removes the surrogate before
psycopg encodes the row.

## Verified end to end

Through mempalace's own backends against a live Postgres + pgvector, a lone surrogate in
`document` or `metadata`:

| backend | result |
|---|---|
| pgvector | aborts (`UnicodeEncodeError: surrogates not allowed`) |
| sqlite_exact | aborts (`UnicodeEncodeError: surrogates not allowed`) |
| chromadb (default) | document-surrogate tolerated (stored as U+FFFD); metadata-surrogate aborts |
| qdrant | not verified (no server available) |

Backend choice changes which inputs are ingestible, but no backend is fully immune.

## Proposed fix

Apply the existing `config.strip_lone_surrogates` (which replaces each lone surrogate with
U+FFFD) to `id`, `document`, and the serialized metadata JSON before insert. Because
`_json_dumps` uses `ensure_ascii=False`, a metadata surrogate stays raw in the serialized
string, so one pass over the JSON cleans it (no dict walk needed):

```python
params = [(
    strip_lone_surrogates(row["id"]),
    strip_lone_surrogates(row["document"]),
    strip_lone_surrogates(_json_dumps(row.get("metadata"))),
    _vector_literal(row["embedding"]),
    row.get("updated_at") or _utcnow(),
) for row in rows]
```

Surrogates are replaced (U+FFFD), not dropped, matching `config.strip_lone_surrogates`.
NUL, by contrast, is stripped (see #1829); the two are independent byte classes on the same
write path and conflict trivially in `upsert_rows`.

## Related

- #1829: the NUL (0x00) sibling on the same `upsert_rows` write path.
- `config.strip_lone_surrogates`: the existing utility (applied to ChromaDB document text)
  that this fix reuses.
- The sqlite_exact gap and ChromaDB's metadata path are the same class of problem and can be
  addressed separately.
- #503 / #1488: lone surrogates over the MCP stdio transport on Windows (cp1252 stdin); a
  different root cause and code path (transport decode, not backend upsert).

## Suggested upstream change

Because at least pgvector and sqlite_exact share this gap, the durable fix is to strip lone
surrogates at the normalizer layer for all backends, matching ChromaDB's tolerance, so
backend choice does not change which inputs are ingestible. This PR fixes the pgvector write
path now, mirroring the #1829 (NUL) precedent.

### PR DESCRIPTION
**Pull request by mvalentsev**
Fixes #1833

## What does this PR do?

pgvector's bulk `upsert_rows` binds `document` (text) and `metadata` (jsonb) straight to
psycopg. A lone UTF-16 surrogate (U+D800-U+DFFF) in transcript content has no UTF-8
encoding, so psycopg raises `UnicodeEncodeError: surrogates not allowed` and the **entire**
`mempalace mine` aborts, leaving every later file unmined.

This applies the existing `config.strip_lone_surrogates` (which replaces each lone surrogate
with U+FFFD) to `id`, `document`, and the serialized metadata JSON in `upsert_rows`:
`strip_lone_surrogates(_json_dumps(row.get("metadata")))`. Because `_json_dumps` uses
`ensure_ascii=False`, a metadata surrogate stays raw in the serialized string, so one pass
over the JSON cleans it without walking the dict. It is the lone-surrogate sibling of the
NUL fix for #1829.

Replace, not drop: a lone surrogate is a broken half-character with a canonical replacement
glyph (U+FFFD), where NUL has none and is stripped. The `id` pass is defensive (real ids are
`sanitize_name`-validated or path-derived plus a hex hash, so a surrogate never reaches a
real id).

Verified end to end against a live Postgres + pgvector through `PgVectorBackend`: before the
fix a surrogate in `document` or `metadata` aborts the mine; after it, the same input ingests
and round-trips with the surrogate stored as U+FFFD. The same lone surrogate also aborts the
sqlite_exact backend and ChromaDB's metadata path; those are out of scope here (this PR
mirrors the pgvector-scoped #1829 precedent) and noted in the issue as follow-ups.

Scope: write path only. The read path is unchanged (rows written after this fix are already
clean, and a surrogate-bearing row could never have been inserted), matching #1829. It
touches `upsert_rows`, so it conflicts trivially with the NUL fix for #1829; whichever lands
second rebases.

## How to test

```
python -m pytest tests/test_pgvector_backend.py -v
```

New tests:
- `test_pgvector_upsert_replaces_lone_surrogates` drives the real `upsert_rows` ->
  `_execute(many=True)` path through a fake psycopg and asserts no lone surrogate survives
  into the bound id/document/metadata parameters (each is UTF-8 encodable) and that content
  round-trips as U+FFFD.
- `test_strip_lone_surrogates_reuses_config_util` pins the `ensure_ascii=False` property that
  lets a single pass over the serialized JSON clean a metadata surrogate. The utility's own
  edge cases (astral chars, surrogate pairs, nesting) live in
  `tests/test_clean_lone_surrogates.py`.

The live-Postgres roundtrip is exercised by the existing
`test_pgvector_live_roundtrip_when_enabled` (gated on `MEMPALACE_PGVECTOR_LIVE_URL`).

## Checklist
- [x] Tests pass (`python -m pytest tests/ -v`)
- [x] No hardcoded paths
- [x] Linter passes (`ruff check .`)

### CODE REVIEW

File: mempalace/backends/pgvector.py [hunk 1]
```diff
@@ -581,9 +624,14 @@ def upsert_rows(self, table: str, rows: list[dict]) -> None:
...
-                row["document"],
-                _json_dumps(row.get("metadata")),
+                # Replace lone surrogates in every text-bound field so none can
+                # abort the insert. Real ids embed only sanitize_name-validated or
+                # path-derived components plus a hex hash, so a surrogate never
+                # reaches a real id; the id pass is defense-in-depth and does not
+                # rewrite the ON CONFLICT key in practice.
+                _strip_lone_surrogates(row["id"]),
+                _strip_lone_surrogates(row["document"]),
+                _json_dumps(_strip_lone_surrogates(row.get("metadata"))),
```
[the remark this answers is not in the document]
Comment (mvalentsev) [reply mempalace/backends/pgvector.py]: Adopted in 765f5fc. Applied `strip_lone_surrogates` to `id`, `document`, and the serialized metadata JSON (`strip_lone_surrogates(_json_dumps(row.get("metadata")))`), and removed the recursive helper.

This is correct because `_json_dumps` uses `ensure_ascii=False`, so a lone surrogate in metadata stays raw in the serialized string and a single pass cleans all three fields. (For NUL the pre-serialization strip is still required, since `json.dumps` escapes NUL to a unicode escape that a post-serialization pass would not catch.)

File: tests/test_pgvector_backend.py [hunk 2]
```diff
@@ -651,3 +653,122 @@ def fake_connect(dsn):
...
+    assert _strip_lone_surrogates((f"a{hi}b", 1, [f"c{lo}"])) == (f"a{repl}b", 1, [f"c{repl}"])
+    # Two keys differing only by a (distinct) surrogate both map to U+FFFD and
+    # collapse, last wins (documented, harmless: real metadata keys are fixed field
+    # names, never surrogate-distinguished).
+    assert _strip_lone_surrogates({f"a{hi}": 1, f"a{lo}": 2}) == {f"a{repl}": 2}
+    # Non-string scalars pass through unchanged (bool stays bool, not int).
+    assert _strip_lone_surrogates(7) == 7
+    assert _strip_lone_surrogates(3.5) == 3.5
+    assert _strip_lone_surrogates(True) is True
+    assert _strip_lone_surrogates(None) is None
```
[the remark this answers is not in the document]
Comment (mvalentsev) [reply tests/test_pgvector_backend.py]: Removed the helper test in 765f5fc along with the helper. The pgvector write path is covered end to end by `test_pgvector_upsert_replaces_lone_surrogates`, and the utility's own edge cases by `tests/test_clean_lone_surrogates.py`. A small test (`test_strip_lone_surrogates_reuses_config_util`) pins the `ensure_ascii=False` assumption the one-pass strip depends on.
